src/pytorch_lightning/loops/fit_loop.py

```diff
@@ -150,31 +150,40 @@ def _results(self) -> _ResultCollection:
     @property
     def done(self) -> bool:
         """Evaluates when to leave the loop."""
+        if self.trainer.num_training_batches == 0:
+            rank_zero_info(f"`Trainer.fit` stopped: No training batches.")
+            return True
+
         # TODO(@awaelchli): Move track steps inside training loop and move part of these condition inside training loop
         stop_steps = _is_max_limit_reached(self.epoch_loop.global_step, self.max_steps)
+        if stop_steps:
+            rank_zero_info(f"`Trainer.fit` stopped: `max_steps={self.max_steps!r}` reached.")
+            return True
+
         # `processed` is increased before `on_train_epoch_end`, the hook where checkpoints are typically saved.
         # we use it here because the checkpoint data won't have `completed` increased yet
         stop_epochs = _is_max_limit_reached(self.epoch_progress.current.processed, self.max_epochs)
         if stop_epochs:
             # in case they are not equal, override so `trainer.current_epoch` has the expected value
             self.epoch_progress.current.completed = self.epoch_progress.current.processed
+            rank_zero_info(f"`Trainer.fit` stopped: `max_epochs={self.max_epochs!r}` reached.")
+            return True
 
-        should_stop = False
         if self.trainer.should_stop:
             # early stopping
             met_min_epochs = self.epoch_progress.current.processed >= self.min_epochs if self.min_epochs else True
             met_min_steps = self.epoch_loop.global_step >= self.min_steps if self.min_steps else True
             if met_min_epochs and met_min_steps:
-                should_stop = True
+                self.trainer.should_stop = True
+                rank_zero_debug(f"`Trainer.fit` stopped: `should_stop` was set.")
+                return True
             else:
-                log.info(
-                    "Trainer was signaled to stop but required minimum epochs"
-                    f" ({self.min_epochs}) or minimum steps ({self.min_steps}) has"
-                    " not been met. Training will continue..."
+                rank_zero_info(
+                    f"Trainer was signaled to stop but the required `min_epochs={self.min_epochs!r}` or"
+                    f" `min_steps={self.min_steps!r}` has not been met. Training will continue..."
                 )
-        self.trainer.should_stop = should_stop
-
-        return stop_steps or should_stop or stop_epochs or self.trainer.num_training_batches == 0
+        self.trainer.should_stop = False
```

Comment:
let's say `min_epochs=2` and `max_epochs=10`
user sets `should_stop=True` after epoch 1. Now training doesn't stop since `min_epochs` haven't met, but should it continue until 10 epochs, or should be stopped after the 2nd epoch once `min_epochs` have met?

asking because we are setting `should_stop=False` here i.e overriding the user's decision.

Reply:
If I'm not mistaken, the changes in this PR do not change the logic regarding that decision.

Now, how it works is that whoever requested to stop will need to re-request it in the next iteration/epoch/whatever. I remember looking into this a while back and there were tests that require this behaviour.

Reply:
The problem is that the inner `TrainingEpochLoop` also checks for `should_stop` in it's `done`. So to do that change, we would need to check for `min_*` in all inner loops that check the attribute

Reply:
so currently the behavior is?

if someone sets `should_stop` in the middle of some epoch, then it breaks that epoch loop and continues running the next epochs completely until `min_epochs` is met.
```
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
>>>>>>>>>>>>>>> (should_stop=True, epoch breaks)
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> (min_epochs)
```

should we add these checks there too to ensure that epoch is completed at least if min_condition is not met?

Reply:
Personally, I would say no (keep as it is). But this is a design question so I don't see a right or wrong answer. Feel free to open an issue to round up the choices and their impact.

Reply:
done! #13789